File: python/blind75/arrays/best-time-stock.py
class Solution:
  def maxProfit(self, prices) -> int:
    largest = 0
    for i, p in enumerate(prices):
      for p2 in prices[i:]:
        d = p2 - p
        # print(d)
        if (d > largest):
          largest = d
    return largest
  def maxProfit2Pointer(self, prices) -> int:
    l, r = 0, 1
    maxi = 0
    while r < len(prices):
      if prices[l] < prices[r]:
        curr = prices[r] - prices[l]
        # print(l, r)
        maxi = max(maxi, curr)    
      else:
        l = r
      r += 1
    return maxi


sol = Solution()
prices = [7,3,5,1,6,4]
# res = sol.maxProfit(prices)
# print(res)
res2 = sol.maxProfit2Pointer(prices)
print(res2)

File: node/best-time-stock.js
var maxProfit = function(p) {
  let l = 0
  let r = 1
  let maxi = 0

  while(r < p.length) {
    if (p[l] < p[r]) {
      c = p[r] - p[l]
      maxi = Math.max(maxi, c)
    } else {
      l = r
    }
    r += 1
  } 
  return maxi
};

const prices = [7,3,5,1,6,4]
const r = maxProfit(prices)
console.log(r)